src/server/routes/consents.ts

```diff
@@ -127,198 +129,317 @@ const getUserNewsletterSubscriptions = async ({
+			newslettersOnPage: ALL_NEWSLETTER_IDS,
+			ip,
+			sc_gu_u,
+			accessToken,
+			request_id,
+		});
+
+		// get a list of newsletters to update that have changed from the users current subscription
+		// if they have changed then set them to subscribe/unsubscribe
+		const newsletterSubscriptionsToUpdate =
+			newslettersSubscriptionsFromFormBody(body).filter((newSubscription) => {
+				// find current user subscription status for a newsletter
+				const currentSubscription = userNewsletterSubscriptions.find(
+					({ id: userNewsletterId }) => userNewsletterId === newSubscription.id,
+				);
+
+				// check if a subscription exists
+				if (currentSubscription) {
+					if (
+						// previously subscribed AND now wants to unsubscribe
+						(currentSubscription.subscribed && !newSubscription.subscribed) ||
+						// OR previously not subscribed AND wants to subscribe
+						(!currentSubscription.subscribed && newSubscription.subscribed)
+					) {
+						// then include in newsletterSubscriptionsToUpdate
+						return true;
+					}
+				}
 
-			return {
-				page: 'review',
-				consents: await getUserConsentsForPage({
-					pageConsents: ALL_CONSENT,
-					ip,
-					sc_gu_u,
-					request_id,
-					accessToken,
-				}),
-				newsletters: await getUserNewsletterSubscriptions({
-					newslettersOnPage: NewsletterMap.get(geo) as string[],
-					ip,
-					sc_gu_u,
-					request_id,
-					accessToken,
-				}),
-			};
-		},
-	},
-];
+				// otherwise don't include in the update
+				return false;
+			});
 
-router.get('/consents', loginMiddlewareOAuth, (_: Request, res: Response) => {
-	const url = addQueryParamsToPath(
-		`${consentPages[0].path}`,
-		res.locals.queryParams,
-	);
+		await patchNewsletters({
+			ip,
+			sc_gu_u,
+			accessToken,
+			payload: newsletterSubscriptionsToUpdate,
+			request_id,
+		});
 
-	res.redirect(303, url);
-});
+		const consents = (ConsentsOnNewslettersPageMap.get(geo) as string[]).map(
+			(id) => ({
+				id,
+				consented: getConsentValueFromRequestBody(id, body),
+			}),
+		);
+
+		await patchConsents({
+			ip,
+			sc_gu_u,
+			accessToken,
+			payload: consents,
+			request_id,
+		});
+	},
+};
+export class ConsentPages {
+	pages: ConsentPage[];
+	//@AB_TEST: 3 Stage Registration Flow Test
+	inSimplifyRegistrationFlowTest = () => {
+		const isInABTestVariant = this.ab.isUserInVariant(
+			abSimplifyRegistrationFlowTest.id,
+			abSimplifyRegistrationFlowTest.variants[0].id,
+		);
+		return isInABTestVariant;
+	};
+
+	threeStageAbTestPages = [OUR_CONTENT, YOUR_DATA(OUR_CONTENT.page), REVIEW];
+	originalPages = [
+		COMMUNICATION,
+		NEWSLETTERS,
+		YOUR_DATA(NEWSLETTERS.page),
+		REVIEW,
+	];
+	constructor(readonly ab: ABTestAPI) {
+		/***
+		  If user is in the AB test bucket, show them the 3 step flow,
+		  otherwise show the original 4 step flow
+		***/
+		if (this.inSimplifyRegistrationFlowTest()) {
+			this.pages = this.threeStageAbTestPages;
+		} else this.pages = this.originalPages;
+	}
+}
```

Comment:
a class seems a bit overkill for this functionality, and is hard to reason about, would it be easier to turn this into a function instead which returns the pages?

```suggestion
export const getConsentPages = (abTestApi: ABTestAPI): SomePagesType => {
  const isInABTestVariant = abTestApi.isUserInVariant(
			abSimplifyRegistrationFlowTest.id,
			abSimplifyRegistrationFlowTest.variants[0].id,
	);
	
	if (isInABTestVariant) {
	  return [OUR_CONTENT, YOUR_DATA(OUR_CONTENT.page), REVIEW]
	}
	return [
		COMMUNICATION,
		NEWSLETTERS,
		YOUR_DATA(NEWSLETTERS.page),
		REVIEW,
	]
}
```

or something along those lines?

Reply:
Yeah that's nicer, thanks